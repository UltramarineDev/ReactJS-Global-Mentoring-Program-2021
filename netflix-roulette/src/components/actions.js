import * as constants from './constants';

const getMoviesSuccessAction = movies => 
  ({ type: constants.GET_MOVIES, payload: movies });

const getMovieSuccessAction = movie => 
  ({ type: constants.GET_MOVIE, payload: movie });

export const getMoviesAction = (filter = undefined, sortBy = constants.DEFAULT_SORT_BY) => async (dispatch) => {
  let url = `${constants.URL}?sortBy=${sortBy}&sortOrder=${constants.DEFAULT_SORT_ORDER}&limit=${constants.DEFAULT_LIMIT}`;

  if (filter) {
    url = `${url}&filter=${filter}`;
  }
  const response = await fetch(url).then(res => res.json());
  dispatch(getMoviesSuccessAction(response.data));
};

export const getMovieAction = (id) => async (dispatch) => {
   const response = await fetch(`${constants.URL}/${id}`).then(res => res.json());
   dispatch(getMovieSuccessAction(response));
};

export const createMovieAction = (input) => async (dispatch) => {
  const response = await fetch(`${constants.URL}`, {
      body: input,
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
    });
};

export const updateMovieAction = (input) => async (dispatch) => {
  const response = await fetch(`${constants.URL}`, {
      body: input,
      method: 'PUT',
      headers: {
        'content-type': 'application/json',
      },
    });
  };
