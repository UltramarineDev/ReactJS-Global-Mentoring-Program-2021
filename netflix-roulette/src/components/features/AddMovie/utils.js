export const getInitialValues = (movie = {}) => ({
  title: movie.title || '',
  releaseDate: movie.release_date || '',
  movieUrl: movie.poster_path || '',
  genre: movie.genres ? movie.genres[0] : '',
  overview: movie.overview || '',
  runtime: movie.runtime || '',
});

export const buildAddMovieInput = (values) => {
  const { title, releaseDate, movieUrl, genre, overview, runtime } = values;
  return {
    title,
    release_date: releaseDate,
    poster_path: movieUrl,
    genres: [genre],
    overview,
    runtime: runtime || 0,
  };
};
