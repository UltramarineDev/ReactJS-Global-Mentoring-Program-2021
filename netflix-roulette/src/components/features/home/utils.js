import { TABS, SORT_OPTIONS, SORT_OPTIONS_LABELS, FILTER_TAGS } from './constants';

export const getTabs = () => ([
  {
    label: 'ALL',
    id: TABS.ALL,
  },
  {
    label: 'DOCUMENTARY',
    id: TABS.DOCUMENTARY,
  },
  {
    label: 'COMEDY',
    id: TABS.COMEDY,
  },
  {
    label: 'HORROR',
    id: TABS.HORROR,
  },
  {
    label: 'CRIME',
    id: TABS.CRIME,
  },
]);

export const getFilteredMovies = (movies, activeTab) => {
  const filterTagList = FILTER_TAGS[activeTab];
  
  if (!movies || movies.length === 0 || !filterTagList) {
    return movies;
  }

  let filteredMovies = movies.filter(({ genres }) => {
    if(!genres || genres.length === 0) {
      return false;
    }

    let matchFound = false;

    genres.forEach((genre) => {
      const areTagsContainsGenre = filterTagList.some((filterTag) => filterTag === genre);
      if (areTagsContainsGenre) {
        matchFound = true;
        return matchFound;
      }
    });

    return matchFound;
  });

  return filteredMovies;
};

export const getOptions = (options, labels) => {
  const resultIds = Object.values(options);

  return resultIds.map(resultId => ({
    id: resultId,
    name: labels[resultId]
  }))
};

//TODO: implement ui sorting in scope of module 4
export const getSortedMovies = (movies, value) => {
  return movies;
};
