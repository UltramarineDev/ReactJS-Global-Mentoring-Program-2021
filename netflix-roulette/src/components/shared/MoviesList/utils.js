import { tabs, filterTags } from './constants';
import { wordings } from '/src/locales/wordings';

export const getTabs = () => ([
  {
    label: wordings.all,
    id: tabs.ALL,
  },
  {
    label: wordings.documentary,
    id: tabs.DOCUMENTARY,
  },
  {
    label: wordings.comedy,
    id: tabs.COMEDY,
  },
  {
    label: wordings.horror,
    id: tabs.HORROR,
  },
  {
    label: wordings.crime,
    id: tabs.CRIME,
  },
]);

export const getOptions = (options, labels) => {
  const resultIds = Object.values(options);

  return resultIds.map(resultId => ({
    id: resultId,
    name: labels[resultId]
  }))
};

//TODO: implement ui sorting in scope of module 4
export const getSortedMovies = (movies, value) => {
  return movies;
};
