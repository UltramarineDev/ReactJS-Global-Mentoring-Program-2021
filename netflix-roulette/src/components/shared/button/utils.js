import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
//TODO: update BUTTON_ICONS
import { BUTTON_ICONS } from '/src/components/shared/constants';

export const getButtonIcon = (icon) => {
  switch (icon) {
    case BUTTON_ICONS.ADD:
      return <FontAwesomeIcon icon="plus" />;
    default: 
      return undefined;
  }
};
