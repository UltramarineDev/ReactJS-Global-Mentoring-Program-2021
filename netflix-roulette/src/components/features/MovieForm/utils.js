import { genres } from 'components/shared/MoviesList/constants';
import { wordings } from 'locales/wordings';

export const getGenres = () => ([
  {
    name: wordings.documentary,
    id: genres.DOCUMENTARY,
  },
  {
    name: wordings.comedy,
    id: genres.COMEDY,
  },
  {
    name: wordings.horror,
    id: genres.HORROR,
  },
  {
    name: wordings.crime,
    id: genres.CRIME,
  },
  {
    name: wordings.action,
    id: genres.ACTION,
  },
]);
