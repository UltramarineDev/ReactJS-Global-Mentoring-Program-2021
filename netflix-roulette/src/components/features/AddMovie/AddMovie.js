import React from 'react';

import MovieForm from '../MovieForm/MovieForm';
import { wordings } from '/src/locales/wordings';
import styles from './AddMovie.module.scss';

const AddMovieForm = () => (
  <div className={styles.addMovieWrapper}>
    <div className={styles.title}>{wordings.add_movie}</div>
    <MovieForm />
  </div>
);

export default AddMovieForm;
