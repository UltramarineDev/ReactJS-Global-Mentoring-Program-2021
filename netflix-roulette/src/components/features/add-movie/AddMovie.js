import React from 'react';

import MovieForm from '../movie-form';

import styles from './AddMovie.module.scss';

const AddMovieFormComponent = () => (
  <>
    <div className={styles.addMovieWrapper}>
      <div className={styles.title}>ADD MOVIE</div>
      <MovieForm />
    </div>
  </>
);

export default AddMovieFormComponent;
