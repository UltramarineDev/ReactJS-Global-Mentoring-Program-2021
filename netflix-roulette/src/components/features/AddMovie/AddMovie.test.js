import React from 'react';
import { fireEvent, render, screen, cleanup } from '@testing-library/react';

import { addMovieAction } from 'actions';

import { buildAddMovieInput, getInitialValues } from './utils';
import AddMovie from './AddMovie';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch,
}));

describe('Add movie tests', () => {
  afterEach(cleanup);
  it('buildAddMovieInput', () => {
    const values = {
      title: 'title',
      releaseDate: 'date',
      movieUrl: 'url',
      genre: 'comedy',
    };

    const expected = {
      title: 'title',
      release_date: values.releaseDate,
      poster_path: values.movieUrl,
      genres: [values.genre],
      runtime: 0,
      overview: undefined,
    };
    expect(buildAddMovieInput(values)).toEqual(expected);
  });

  it('getInitialValues', () => {
    const movie = {
      release_date: 'date',
      poster_path: 'url',
      genres: 'comedy',
      overview: 'text',
      runtime: 'time',
    };

    const expected = {
      title: '',
      releaseDate: movie.release_date,
      movieUrl: movie.poster_path,
      genre: movie.genres[0],
      overview: movie.overview,
      runtime: movie.runtime,
    };

    expect(getInitialValues(movie)).toEqual(expected);
  });

  it('renders AddMovie', () => {
    const onSave = jest.fn();
    const { asFragment } = render(<AddMovie onSave={onSave} />);
    expect(asFragment(<AddMovie onSave={onSave} />)).toMatchSnapshot();
  });

  // it('calls onSave callback handler', () => {
  //   const onSave = jest.fn();

  //   const { getByText } = render(<AddMovie onSave={onSave} />);
  //   const button = getByText(/save/i);

  //   fireEvent.click(button);

  //   expect(onSave).toHaveBeenCalledTimes(1);
  //   // expect(buildAddMovieInput).toHaveBeenCalledTimes(1);
  //   // expect(addMovieAction).toHaveBeenCalledTimes(1);
  // });
});
