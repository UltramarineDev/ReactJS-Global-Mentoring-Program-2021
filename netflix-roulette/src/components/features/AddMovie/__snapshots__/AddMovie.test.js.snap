// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add movie tests renders AddMovie 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <div
      class="title"
    >
      ADD MOVIE
    </div>
    <p
      class="label"
    >
      TITLE
    </p>
    <input
      id="title"
      name="title"
      placeholder="Title here"
      type="text"
      value=""
    />
    <p
      class="label"
    >
      RELEASE DATE
    </p>
    <input
      id="releaseDate"
      placeholder="Select Date"
      type="date"
      value=""
    />
    <p
      class="label"
    >
      MOVIE URL
    </p>
    <input
      id="movieUrl"
      placeholder="Movie URL here"
      type="text"
      value=""
    />
    <p
      class="label"
    >
      GENRE
    </p>
    <div
      class="dropdown"
    >
      <div
        class="dropdown"
      >
        <a
          data-testid="set_options_dropdown_label"
        >
          Select Genre
        </a>
        <div
          class="icon"
        />
      </div>
    </div>
    <p
      class="label"
    >
      OVERVIEW
    </p>
    <input
      id="overview"
      placeholder="Overview here"
      type="text"
      value=""
    />
    <p
      class="label"
    >
      RUNTIME
    </p>
    <input
      id="runtime"
      placeholder="Runtime here"
      type="text"
      value=""
    />
    <button
      class="searchButton smallSize"
      data-testid="save_form"
      type="submit"
    >
      <span
        class="labelSearch"
      >
        SAVE
      </span>
    </button>
  </form>
</DocumentFragment>
`;
